Readme.md

# scraper

  A simpler API for [node-phantom](https://github.com/sgentle/phantomjs-node) web scraping.

## Installation

    $ npm install ivolo/scraper

## Example

### Create a Scraper

```js
var Scraper = require('scraper');

Scraper(function (err, scraper) {
  // creates a `Scraper` plantom instance on a random port
});
```

### Open a Page

```js
scraper.page(function (err, page) {
  // open a page with disguised headers ..
  page.close();
});
```

### Open a Ready Page

```js
scraper.readyPage(url, function (err, page) {
  // do something with the page
  page.close();
});
```

### Get a Ready Page's HTML

```js
scraper.html(url, function (err, page, html) {
  // do something with the html
  page.close();
});
```

## API

### Command Line

```
Usage: scrape url

Options:

  -h, --help     output usage information
  -v, --version  output the version number
  -e, --eval     evaluate the code

Examples:

  # print out html
  $ scrape https://google.com

  # evalute javascript
  $ scrape https://google.com "document.title"

```

```
$ scrape https://google.com "document.title"
'Google'
```

### Node.JS

### Scraper([options], callback)

  Creates a Scraper instance, with defaulted `options`:

```
{
  port: 12300 + Math.floor(Math.random() * 10000), // defaults to a random port
  flags: ['--load-images=no'],
  headers: { // disguise headers
    'user-agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_8_4) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/28.0.1500.71 Safari/537.36',
    'accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8',
    'accept-language' :'en-US,en;q=0.8',
    'cache-control' :'max-age=0'
  }
}
```

### Scraper

  A `Scraper` instance is returned by the `Scraper(callback)` method.

#### Scraper.phantom

  Get the phantom `instance`. An example to clear the cookies using the original node-phantom API:

```js
var Scraper = require('scraper');

Scraper(function (err, scraper) {
  scraper.phantom.clearCookies(function (err) {
    // cookies cleared
  });
});
```

#### Scraper.page([options], callback)

  Open a page using the disguised headers, with defaulted `options`: 

```js
{
  headers: { .. }
}
```

#### Scraper.readyPage(url, [options], callback)

  Open a page using the disguised headers, load the page at the `url`, and wait for `document.readyState === 'complete'` before returning the page. Uses defaulted `options`: 

```js
{
  checkEvery: 500, // checks document.readyState every so many ms until its ready
  after: 1000, // will wait this many ms after document.readyState to let javascript alter the page
  headers: { .. }
}
```

#### Scraper.html(url, [options], callback)

  Open a ready page at `url`, and return the html. Uses same `options` defaults as `Scraper.readyPage`.


## License

```
WWWWWW||WWWWWW
 W W W||W W W
      ||
    ( OO )__________
     /  |           \
    /o o|    MIT     \
    \___/||_||__||_|| *
         || ||  || ||
        _||_|| _||_||
       (__|__|(__|__|
```


History.md
0.1.1 - January 14, 2014
-------------------------
- add silencing handlers

0.1.0 - January 11, 2014
-------------------------
- rename cli to `scrape`

0.0.1 - December 18, 2013
-------------------------
:sparkles:

test/index.js

var assert = require('assert');
var Scraper = require('..');

describe('scraper', function () {
  this.timeout(5000); // booting up phantom can take a second

  it('should be able to create a new phantom instance', function (done) {
    Scraper(function (err, scraper) {
      if (err) return done(err);
      assert(scraper);
      assert(scraper.phantom);
      done();
    });
  });

  it('should be able to open a page', function (done) {
    Scraper(function (err, scraper) {
      if (err) return done(err);
      scraper.page(function (err, page) {
        if (err) return done(err);
        page.close();
        done();
      });
    });
  });

  it('should be able to open a ready page', function (done) {
    Scraper(function (err, scraper) {
      if (err) return done(err);
      scraper.page(function (err, page) {
        if (err) return done(err);
        assert(page);
        page.close();
        done();
      });
    });
  });

  it('should be able to open a ready page', function (done) {
    this.timeout(30000); // need to wait extra for page ready
    Scraper(function (err, scraper) {
      if (err) return done(err);
      scraper.readyPage('https://google.com', function (err, page) {
        if (err) return done(err);
        assert(page);
        page.close();
        done();
      });
    });
  });

  it('should be able to get a pages html', function (done) {
    this.timeout(30000); // need to wait extra for page ready
    Scraper(function (err, scraper) {
      if (err) return done(err);
      scraper.html('https://bing.com', function (err, page, html) {
        if (err) return done(err);
        assert(page);
        page.close();
        assert(html);
        assert(html.length > 0);
        done();
      });
    });
  });
});

index.js

var debug = require('debug')('scraper');
var defaults = require('defaults');
var Emitter = require('events').EventEmitter;
var inherit = require('util').inherits;
var phantom = require('phantom');

/**
 * Expose `create`.
 */

module.exports = create;

/**
 * Creates a `Scraper` instance.
 *
 * @param {Object} options
 * @param {Function} callback
 */

function create (options, callback) {
  if (typeof options === 'function') {
    callback = options;
    options = {};
  }

  options = defaults(options, {
    port: 12300 + Math.floor(Math.random() * 10000), // defaults to a random port
    flags: ['--load-images=no'],
    headers: { // disguise headers
      'user-agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_8_4) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/28.0.1500.71 Safari/537.36',
      'accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8',
      'accept-language' :'en-US,en;q=0.8',
      'cache-control'   :'max-age=0'
    }
  });

  debug('creating phantom instance at port %d ..', options.port);
  phantom.create(options.flags, options, function (instance) {
    debug('created phantom instance at port %d', options.port);
    return callback(null, new Scraper(instance, options));
  });
}

/**
 * Create a new `Scraper` instance.
 *
 * @param {Phantom} phantom
 */

function Scraper (phantom, options) {
  if (!(this instanceof Scraper)) return new Scraper(phantom, options);
  this.phantom = phantom;
  this.options = options;
}

/**
 * Inherit from `Emitter`.
 */

inherit(Scraper, Emitter);

/**
 * Open a page using the disguised headers.
 *
 * @param {Object} options
 * @param {Function} callback
 */

Scraper.prototype.page = function (options, callback) {
  if (typeof options === 'function') {
    callback = options;
    options = {};
  }

  options = defaults(options, this.options);

  debug('creating disguised phantom page ..');

  this.phantom.createPage(function (page) {
    disguise(page, options.headers);
    debug('created disguised phantom page');
    return callback(null, page);
  });
};

/**
 * Open a page using the disguised headers, load the
 * page at the `url`, and wait for
 * `document.readyState === 'complete'`
 * before returning the page.
 *
 * @param {String} url
 * @param {Object} options
 * @param {Function} callback
 */

Scraper.prototype.readyPage = function (url, options, callback) {
  if (typeof options === 'function') {
    callback = options;
    options = {};
  }

  this.page(options, function (err, page) {
    if (err) return callback(err);
    page.open(url, function (status) {
      debug('page %s opened with status %s', url, status);
      if (status !== 'success') {
        return callback(new Error('Opening page' + url +
          'resulted in status ' + status));
      }

      waitForReady(page, function (err) {
        return callback(err, page);
      });
    });
  });
};

/**
 * Open a ready page at `url` and return the html.
 *
 * @param {String} url
 * @param {Object} options
 * @param {Function} callback
 */

Scraper.prototype.html = function (url, options, callback) {
  if (typeof options === 'function') {
    callback = options;
    options = {};
  }

  this.readyPage(url, options, function (err, page) {
    if (err) return callback(err);
    getPageHtml(page, function (err, html) {
      return callback(err, page, html);
    });
  });
};

/**
 * Return a map copy with all the keys lowercased.
 *
 * @param {Map} map
 * @return {Object}
 */

function lowercase (map) {
  var result = {};
  Object.keys(map).forEach(function (key) {
    result[key.toLowerCase()] = map[key];
  });
  return result;
}

/**
 * Disguise a `page` with custom `headers`.
 *
 * @param {Page} page
 * @param {Object} headers
 */

function disguise (page, headers) {
  var lowercased = lowercase(headers);
  var userAgent = lowercased['user-agent'];
  if (userAgent) page.set('settings.userAgent', userAgent);
  page.set('customHeaders', lowercased);
}

/**
 * Waits until page's document.readyState === complete.
 *
 * @param {Page} page
 * @param {Object} options
 *   @param {Number} checkEvery
 *   @param {Number} after
 * @param {Function} callback
 */

function waitForReady (page, options, callback) {
  if (typeof options === 'function') {
    callback = options;
    options = {};
  }

  options = defaults(options, {
    checkEvery: 500, // to retry document.readyState
    after: 1000 // wait for javascript/ajax
  });

  debug('waiting for page document.readyState === complete ..');
  page.evaluate(
    function () { return document.readyState; },
    function (result) {
      debug('page is document ready, waiting for javascript/ajax timeout ..');
      if (result === 'complete') {
        return setTimeout(function () {
          debug('page is "ready"');
          return callback();
        }, options.after);
      }
      else {
        setTimeout(function () {
          waitForReady(page, callback);
        }, options.checkEvery);
      }
    }
  );
}

/**
 * Get the `page`s html.
 *
 * @param {Page} page
 * @param {Function} callback
 */

function getPageHtml (page, callback) {
  debug('getting page html ..');
  page.evaluate(
    function () {
      return '<html>' + document.head.outerHTML +
      document.body.outerHTML + '</html>';
    },
    function (html) {
      debug('got page html: %d chars', html.length);
      return callback(null, html);
    }
  );
}

/**
 * Add a Phantom standard error handler for silence.
 * https://github.com/ivolo/scraper/issues/2
 *
 * @param {String} message
 */

phantom.stderrHandler = function (message) {
 if(message.match(/(No such method.*socketSentData)|(CoreText performance note)|(WARNING: Method userSpaceScaleFactor in class NSView is deprecated on 10.7 and later.)/)) return;
 console.error(message);
};